packages/block-editor/src/store/reducer.js

```diff
@@ -1282,7 +1282,21 @@ export function lastBlockAttributesChange( state, action ) {
  * @return {boolean} Updated state.
  */
 export function didAutomaticChange( state, action ) {
-	return action.type === 'MARK_AUTOMATIC_CHANGE';
+	switch ( action.type ) {
+		case 'MARK_AUTOMATIC_CHANGE':
+			return 'pending';
+		case 'MARK_AUTOMATIC_CHANGE_FINAL':
+			if ( state === 'pending' ) {
+				return 'final';
+			}
+
+			return;
+		case 'SELECTION_CHANGE':
+			// As long as the state is not final, ignore any selection changes.
+			if ( state !== 'final' ) {
+				return state;
+			}
+	}
```

Comment:
* The function signature should be updated so the `@return` type moves from bool to string. The name should also be changed from the `did*` boolean form.
* When action type is `SELECTION_CHANGE` but `state` is `'final'`, the reducer will implicitly return `undefined`. That value might be fine (though we might prefer `null`? no strong opinion here), but let's make it explicit by adding a fallback return statement.

Reply:
@mcsf What name would you suggest for the reducer? For any other action we also explicitly return `undefined`. Personally I like it to be `undefined`. I'm fine with adding an explicit return statement for the action we're handling.

Reply:
The name could be as plain as `automaticChangeStatus`, since that's a reflection of what it contains.

> For any other action we also explicitly return `undefined`.

`undefined` is fine with me, as long as explicit.

Meanwhile, this reducer is a bit particular. For most reducers, the default behaviour is to keep state unchanged. That is, unless the reducer receives a particular action that matters, we always return `state` in the end. This isn't the case with `didAutomaticChange`, though, because the default behaviour is now to reset to `undefined`. So I would add that as something explicit too:

```js
switch ( … ) {
 …
}

// Once we're no longer marking changes or changing selection, reset state.
return undefined;
```

Reply:
To me, that was the beauty of of it. By default, a reducer resets any state to `undefined`, which we make use of here. Returning undefined explicitly here seems a bit strange to me (`null` sounds better then). A return statement returns `undefined` by default, so it could be omitted. A function returns `undefined` by default, so it could be omitted (returning nothing at the end of a function seems strange as well).

Reply:
We could still comment in the reducer though, that by default, we reset state instead of keeping state.

Reply:
As long as there's something that reveals intention, even if just a comment, that's what matters to me.